backend/src/main/java/pl/edu/agh/niebieskiekotki/groups/GroupGenerator.java

```diff
@@ -0,0 +1,123 @@
+package pl.edu.agh.niebieskiekotki.groups;
+
+import org.jgrapht.alg.flow.DinicMFImpl;
+import org.jgrapht.graph.DefaultDirectedWeightedGraph;
+import org.jgrapht.graph.DefaultWeightedEdge;
+import org.springframework.stereotype.Component;
+import pl.edu.agh.niebieskiekotki.entitites.Student;
+import pl.edu.agh.niebieskiekotki.entitites.Term;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Component
+public class GroupGenerator {
+    public GenerationOutput generate(Map<Student, List<Term>> studentsTerms, int numGroups){
+        var network = new DefaultDirectedWeightedGraph<>(DefaultWeightedEdge.class);
+        var source = new Object();
+        var sink = new Object();
+
+        network.addVertex(source);
+        network.addVertex(sink);
+
+        Set<Student> allStudents = studentsTerms.keySet();
+
+        for (var student: allStudents){
+            network.addVertex(student);
+            network.addEdge(source, student);
+        }
+
+        Set<Term> allTerms = getAllTerms(studentsTerms);
+        Set<Term> candidateTerms = getCandidateTerms(studentsTerms, allTerms, numGroups);
```

Comment:
Może warto by pobierać o kilku kandydatów dla terminów więcej i liczyć dla każdej kombinacji tych wybranych terminów podział na grupy po czym wybierać ten podział dla tych terminów ktory jest najlepszy? 
Np jeżeli mamy 5 grup to można by pobierać 8 kandydatów i liczyć podział na każdy możliwy wybór 5 z tych 8 terminów.
Można by też uzależnić ilość wybieranych dodatkowo kandydatów od ilości grup, np jeżeli ilość grup jest 5 to spokojnie możemy pobrać dodatkowe trzy terminy i nie zabije to całkiem złożoności, ale jeżeli grup jest 20 to prawdopodobnie wystarczy jeden dodatkowy termin (albo zero), zwłaszcza że będziemy mieli już sporo terminów pomiędzy które będziemy rozdzielać studentów więc tacy dodatkowi kandydaci nie powinni mieć tak dużego znaczenia.
Czyli im mniej grup tym więcej dodatkowych kandydatów byśmy wybierali, ponieważ mamy więcej mocy obliczeniowej "w zapasie".
Idk, powiedz co sądzisz o tym pomyśle

Reply:
Tak, jak najbardziej można tak polepszyć algorytm. Myślałem o tym, ale wolałem zaimplemetować prostszą wersję, żeby to w miarę szybko skończyć

Myślę, że mogę śmiało dodać taki improvement w sprincie 3, jeśli będzie na to czas